components/eamxx/src/share/util/eamxx_time_interpolation.cpp

```diff
@@ -0,0 +1,340 @@
+    const int mm = (time_start - hh*10000)/100;
+    const int ss = (time_start - hh*10000 - mm*100);
+    TimeStamp ts_file_start(YY,MM,DD,hh,mm,ss);
+    if (ii==0) {
+      ts_ref = ts_file_start;
+    }
+    // Gather information about time in this file
+    scorpio::register_file(filename,scorpio::Read);
+    const int ntime = scorpio::get_dimlen(filename,"time");
+    for (int tt=0; tt<ntime; tt++) {
+      auto time_snap = scorpio::read_time_at_index_c2f(filename.c_str(),tt+1);
+      TimeStamp ts_snap = ts_file_start;
+      if (time_snap>0) {
+        ts_snap += (time_snap*86400); // note, time is assumed to be in days.
+      }
+      auto time = ts_snap.seconds_from(ts_ref);
+      map_of_times_to_vector_idx.emplace(time,running_idx);
+      filenames_tmp.push_back(filename);
+      timestamps_tmp.push_back(ts_snap);
+      time_idx_tmp.push_back(tt);
+      ++running_idx;
+    }
+  }
+  // Now that we have gathered all of the timesnaps we can arrange them in order as DataFromFileTriplet objects.
+  // Taking advantage of maps automatically self-sorting by the first arg.
+  for (auto a : map_of_times_to_vector_idx) {
+    auto idx = a.second;
+    DataFromFileTriplet my_trip;
+    my_trip.filename  = filenames_tmp[idx];
+    my_trip.timestamp = timestamps_tmp[idx];
+    my_trip.time_idx  = time_idx_tmp[idx];
+    m_file_data_triplets.push_back(my_trip);
+  }
+  // Finally set the iterator to point to the first triplet.
+  m_triplet_iterator = m_file_data_triplets.begin();
+}	
+/*-----------------------------------------------------------------------------------------------*/
+/* Function to read a new set of data from file using the current iterator pointing to the current
+ * DataFromFileTriplet.
+ */
+void TimeInterpolation::read_data()
+{
+  if (m_triplet_iterator->filename != m_file_data_atm_input.get_filename()) {
+    // Then we need to close this input stream and open a new one
+    m_file_data_atm_input.finalize();
+    ekat::ParameterList input_params;
+    input_params.set("Field Names",m_field_names);
+    input_params.set("Filename",m_triplet_iterator->filename);
+    m_file_data_atm_input = AtmosphereInput(input_params,m_fm_time1);
+  }
+  m_file_data_atm_input.read_variables(m_triplet_iterator->time_idx);
+  m_time1 = m_triplet_iterator->timestamp;
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function to check the current set of interpolation data against a timestamp and, if needed,
+ * update the set of interpolation data to ensure the passed timestamp is within the bounds of
+ * the interpolation data.
+ * Input:
+ *   ts_in - A timestamp that we intend to interpolate onto.
+ */
+void TimeInterpolation::check_and_update_data(const TimeStamp& ts_in)
+{
+  // First check if the passed timestamp is within the bounds of time0 and time1.
+  bool current_data_okay = (ts_in.seconds_from(m_time0) >= 0) and (m_time1.seconds_from(ts_in) >= 0);
```

Comment:
Looks like the first check should always be true (unless we start from the left of the first time interval). I think the first check should be inside an `EKAT_REQUIRE_MSG`, to ensure we are not extrapolating on the left.